tabby-terminal: init at 1.0.230

dtomvan opened the pull request:
An attempt at reviving the effort to package https://tabby.sh/ for NixOS.

- Cleaned up the code
- Updated to latest version
- Used electron headers from `electron.headers` instead of a manual FOD
- Wrote an update script
- Disabled automatic updates because that doesn't work on NixOS and isn't expected anyways

cc @geodic

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

geodic committed:
maintainers: add geodic

(cherry picked from commit 990383a00eea7ca96baf25f3687cd5e17b40143e)
(cherry picked from commit 6fd17cd59ca8249bdf5766197e48001cbe5776db)

dtomvan committed:
tabby-terminal: init at 1.0.230

Co-authored-by: Geodic <th3geodic@proton.me>

geodic commented:
Congrats @dtomvan! I was planning to do this myself, but my other projects were pulling me away. Even if my changes got merged though, I would probably be too busy to maintain this without the automatic update script.

geodic commented on pkgs/by-name/ta/tabby-terminal/package.nix:
```diff
@@ -0,0 +1,210 @@
+    repo = "tabby";
+    tag = "v${version}";
+    hash = "sha256-M9i8q6cDT6XqIFG0G+kkXMlxlEmAsGPZv01/7/q89cw=";
+    leaveDotGit = true;
```
`leaveDotGit` is bad as the hash constantly changes due to new commits, etc. The ideal solution to fix this would be to patch the code running git in the first place. After a little searching, it looks like the offending code is right [here](https://github.com/Eugeny/tabby/blob/61f02d1f851971df043f28677a4b37a4bc7f6871/scripts/vars.mjs#L13). I presume you can use something like `substituteInPlace` to hardcode the version.

ecotechie commented:
Outside of it being actually committed, is there anything holding this up?

dtomvan commented:
Squashed all commits into one, added myself as maintainer, bumped electron, fixed some overridability problems, cleaned stuff up. Should be good to go now. The patch I added is because `node-abi` doesn't know that electron 41 exists, so that's temporary until tabby gets a new release I suppose.

<details>

<summary>The `node-abi` error I'm talking about</summary>

```
Error: Could not detect abi for version 41.2.0 and runtime electron.  Updating "node-abi" might help solve this issue if it is a new release of electron
    at getAbi (file:///build/source/node_modules/node-abi/index.js:32:9)
    at get ABI (file:///build/source/node_modules/@electron/rebuild/lib/rebuild.js:79:31)
    at PrebuildInstall.prebuiltModuleExists (file:///build/source/node_modules/@electron/rebuild/lib/module-type/prebuild-install.js:54:153)
    at ModuleRebuilder.prebuildInstallNativeModuleExists (file:///build/source/node_modules/@electron/rebuild/lib/module-rebuilder.js:48:37)
    at Rebuilder.rebuildModuleAt (file:///build/source/node_modules/@electron/rebuild/lib/rebuild.js:136:35)
    at file:///build/source/node_modules/@electron/rebuild/lib/rebuild.js:93:43
    at Rebuilder.rebuild (file:///build/source/node_modules/@electron/rebuild/lib/rebuild.js:101:23)
    at async file:///build/source/node_modules/@electron/rebuild/lib/cli.js:130:9
```

</details>

dtomvan commented:
I tried to reduce the output size (~1GiB) by using `electron-builder`, but the app errored:

```
$ ./result/bin/tabby
[...]
Error: Cannot find module 'v8-compile-cache'
Require stack:
- /nix/store/inr8fqjzd5ddy8c41gaj9rb2kffjvf11-tabby-terminal-1.0.230/share/tabby/resources/app.asar/dist/main.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1463:15)
    at s._resolveFilename (node:electron/js2c/browser_init:2:140033)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1073:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1078:22)
    at Module._load (node:internal/modules/cjs/loader:1249:25)
    at c._load (node:electron/js2c/node_init:2:18047)
    at wrapModuleLoad (node:internal/modules/cjs/loader:262:19)
    at Module.require (node:internal/modules/cjs/loader:1563:12)
    at require (node:internal/modules/helpers:152:16)
    at 4997 (/nix/store/inr8fqjzd5ddy8c41gaj9rb2kffjvf11-tabby-terminal-1.0.230/share/tabby/resources/app.asar/dist/main.js:2:490756)
```
